backstory/src/main/java/com/google/sps/story/data/NLServiceClient.java

```diff
@@ -0,0 +1,185 @@
+import com.google.cloud.language.v1.PartOfSpeech.Form;
+import com.google.cloud.language.v1.PartOfSpeech.Proper;
+import com.google.cloud.language.v1.PartOfSpeech.Tag;
+import com.google.cloud.language.v1.Token;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Service client for Cloud Natural Language API
+ */
+public class NLServiceClient {
+  /** holds the language service client instance for this service client */
+  private LanguageServiceClient client;
+
+  /**
+   * Constructs a default object of client service class.
+   *
+   * @throws IOException if an instance of LanguageServiceClient could not be created
+   */
+  public NLServiceClient() throws IOException {
+    this(LanguageServiceClient.create());
+  }
+
+  /**
+   * Constructs a NLServiceClient class with a given LanguageServiceClient
+   * (mainly to be used for testing)
+   *
+   * @param client a given LanguageServiceClient to use to access Natural Language API
+   * @throws IllegalArgumentException if client is null
+   */
+  public NLServiceClient(LanguageServiceClient client) throws IllegalArgumentException {
+    if (client == null) {
+      throw new IllegalArgumentException("Client cannot be null");
+    }
+
+    this.client = client;
+  }
+
+  /**
+   * Takes a list of words and sorts them by word type.
+   * Returns a map which has word types as keys and then words
+   * in a list linked to their relevant word type (e.g. "Mary" would
+   * be sorted into the PROPER_NOUN list and "calm" would be sorted
+   * into the adjective list).
+   *
+   * @param words the words to sort by word type
+   * @return a map with lists of words sorted by their word type and with their
+   *    word type as a key
+   */
+  public Map<WordType, List<String>> sortByWordType(List<String> words) {
+    Map<WordType, List<String>> map = new HashMap<WordType, List<String>>();
+
+    List<String> singleWords = new ArrayList<String>();
+
+    // remove multi-word inputs first and store as multiword nouns
+    for (String text : words) {
+      if (text.contains(" ")) {
+        if (map.containsKey(WordType.MULTIWORD_NOUN)) {
+          List<String> wordList = map.get(WordType.MULTIWORD_NOUN);
+          wordList.add(text);
+        } else {
+          ArrayList<String> wordList = new ArrayList<String>();
+          wordList.add(text);
+          map.put(WordType.MULTIWORD_NOUN, wordList);
+        }
+      } else {
+        singleWords.add(text);
+      }
+    }
+
+    for (String word : singleWords) {
+      Document doc = buildDocumentFromText(word);
+
+      AnalyzeSyntaxResponse response = client.analyzeSyntax(doc);
+
+      Token token = response.getTokens(0);
+      PartOfSpeech partOfSpeech = token.getPartOfSpeech();
+
+      WordType type = WordType.UNUSABLE;
+
+      switch (partOfSpeech.getTag()) {
+        case VERB:
+          if (isGerund(word)) {
```

Comment:
Are gerunds be classified as both nouns and verbs by the API?

Reply:
Gerunds can be classified as both nouns and verbs (it depends on the context they appear in). They're more likely to be classified as nouns when they appear solo like they will at first (but sometimes they have been classified as verbs initially), which is why there's an isGerund() method, which has a more foolproof way to check if it's a gerund.

Reply:
Done